runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -3549,6 +3549,38 @@ void genInstanceOfOrCheckCastSuperClassTest(TR::Node *node, TR::Register *condRe
    {
    TR::Compilation *comp = cg->comp();
 
+   // When castClass is an unknown at compile time check if castClass is an interface or an array.
+   // If so skip superClassTest.
+   bool dynamicCastClass = castClassDepth == -1;
+   if ( dynamicCastClass )
+      {
+      TR::Register *scratchRegister1 = srm->findOrCreateScratchRegister();
+      TR::Register *scratchRegister2 = srm->findOrCreateScratchRegister();
+      TR_ASSERT(node->getSecondChild()->getOpCodeValue() != TR::loadaddr,
+            "genInstanceOfOrCheckCastSuperClassTest: castClassDepth == -1 is not supported for a loadaddr castClass");
+
+      generateTrg1MemInstruction(cg, TR::InstOpCode::Op_load, node, scratchRegister1, TR::MemoryReference::createWithDisplacement(cg, castClassReg, offsetof(J9Class, romClass), TR::Compiler->om.sizeofReferenceAddress()));
+      generateTrg1MemInstruction(cg, TR::InstOpCode::lwz, node, scratchRegister1, TR::MemoryReference::createWithDisplacement(cg, scratchRegister1, offsetof(J9ROMClass, modifiers), 4));
+
+      // Array check
+      if (J9AccClassArray <= UPPER_IMMED && J9AccClassArray >= LOWER_IMMED)
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::andi_r, node, scratchRegister2, scratchRegister1, condReg, J9AccClassArray);
+      else
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::andis_r, node, scratchRegister2, scratchRegister2, condReg, J9AccClassArray >> 16);
+      // At this point cr0[eq] will be set if this is not an array.
+      generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, node, falseLabel, condReg);
+      // Interface check
+      if (J9AccInterface <= UPPER_IMMED && J9AccInterface >= LOWER_IMMED)
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::andi_r, node, scratchRegister2, scratchRegister1, condReg, J9AccInterface);
+      else
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::andis_r, node, scratchRegister2, scratchRegister2, condReg, J9AccInterface >> 16);
+      // At this point cr0[eq] will be set if this is not an interface.
+      generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, node, falseLabel, condReg);
+
+      srm->reclaimScratchRegister(scratchRegister1);
+      srm->reclaimScratchRegister(scratchRegister2);
+      }
+
```

Comment:
If `castClassDepth == -1` is true at compile-time do we have to generate code to load it at run-time? The aarch64 codegen has

```
...

 *  @param[in] castClassDepth:                 class depth of the cast class. If -1 is passed, depth is loaded at runtime

...

   if (castClassDepth != -1)
      {
      // castClassDepth is known at compile time
      ...
      }
   else
      {
      // castClassDepth needs to be loaded from castClass
      castClassDepthReg = srm->findOrCreateScratchRegister();
      // load lower 16bit of classDepthAndFlags
      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrhimm, node, castClassDepthReg,
                              TR::MemoryReference::createWithDisplacement(cg, castClassReg, offsetof(J9Class, classDepthAndFlags)));
      generateCompareInstruction(cg, node, instanceClassDepthReg, castClassDepthReg);
      }
```

which loads `castClassReg->classDepthAndFlags` but the Power codegen only has

```
   if (castClassDepth <= UPPER_IMMED && castClassDepth >= LOWER_IMMED)
      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpi4, node, condReg, instanceClassDepthReg, castClassDepth);
   else
      {
      castClassDepthReg = srm->findOrCreateScratchRegister();
      loadConstant(cg, node, castClassDepth, castClassDepthReg);
      generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp4, node, condReg, instanceClassDepthReg, castClassDepthReg);
      }
```

which is generating code to use `castClassDepth` either as an immediate if it fits or via a register otherwise, but not doing anything based on whether it is `-1` or not.

Reply:
Correct, good catch. The current superClassTest inlining code does not handle unknown classDepth and generates useless code. Adding to my WIP list.

Reply:
Fixed in 8000695e619da88a7ecb84ca27cd7e3524988fac